mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+                index=0;
+                for(Object id:idsInBatch) {
+                    if(!results.containsKey(index)) { // No other errors for this id
+                        if(!successfulIds.contains(id)) {
+                            // concurrency errors for this id
+                            results.put(index,Error.get("update",MongoCrudConstants.ERR_CONCURRENT_UPDATE,id.toString()));
+                        }
+                    }
+                    index++;
+                }
+            } finally {
+                if(cursor!=null)
+                    cursor.close();
+            }
+        }
+    }
+
+    /**
+     * Writes a replace query, which is:
+     * <pre>
+     *  {_id:<docId>, "@mongoHidden.docver.0":ObjectId(<originalDocver>)}
+     * </pre>
+     */
+    private DBObject writeReplaceQuery(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        ObjectId top=null;
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null&&!list.isEmpty())
+                top=list.get(0);
+        }
+        BasicDBObject query=new BasicDBObject("_id",doc.get("_id"));
+        query.append(DOCVER_FLD0,top);
+        return query;
+    }
+    
+
+    /**
+     * Returns the @mongoHidden at the root level of the document. Adds one if necessary.
+     */
+    public static DBObject getHidden(DBObject doc,boolean addIfNotFound) {
+        DBObject hidden=(DBObject)doc.get(Translator.HIDDEN_SUB_PATH.toString());
+        if(hidden==null&&addIfNotFound) {
+            doc.put(Translator.HIDDEN_SUB_PATH.toString(),hidden=new BasicDBObject());            
+        }
+        return hidden;
+    }
+
+    public static void overwriteDocVer(DBObject doc,ObjectId docver) {
+        getHidden(doc,true).removeField(DOCVER);
+        setDocVer(doc,docver);
+    }
+    
+    public static void setDocVer(DBObject doc,ObjectId docver) {
+        DBObject hidden=getHidden(doc,true);
+        List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+        if(list==null) {
+            list=new ArrayList<ObjectId>();
+        }
+        list.add(0,docver);
+        hidden.put(DOCVER,list);
+    }
+
+    /**
+     * Removes old docvers from a document
+     */
+    private void cleanupOldDocVer(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null) {
+                List<ObjectId> copy=new ArrayList<>(list.size());
+                long now=docVer.getDate().getTime();
+                for(ObjectId id:list) {
+                    if(!id.equals(docVer)) {
+                        Date d=id.getDate();
+                        if(now-d.getTime()<TOO_OLD) {
+                            copy.add(id);
+                        }
+                    } else {
+                        copy.add(id);
+                    }
+                }
```

Comment:
Maybe we could have both timeout _and_ max array length (keeping old versions if max length not met yet). This would keep most ids possible in case it takes ridiculously long to find concurrent updates.

I might just be being paranoid on that though. Maybe it's enough to just have long enough timeout.